fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReplacePredicate.java

```diff
@@ -0,0 +1,320 @@
+        if (targetExprs.isEmpty()) {
+            return new LinkedHashSet<>(inputs);
+        }
+        Map<Expression, Set<Expression>> exprPredicates = new HashMap<>();
+        for (Expression input : inputs) {
+            if (input.anyMatch(expr -> !((ExpressionTrait) expr).isDeterministic())
+                    || input.getInputSlots().size() != 1) {
+                continue;
+            }
+            input.accept(PredicatesCollector.INSTANCE, exprPredicates);
+        }
+        Set<Expression> inferPredicates = new LinkedHashSet<>(inputs);
+        if (!exprPredicates.isEmpty()) {
+            for (Expression expr : targetExprs) {
+                if (expr instanceof Literal) {
+                    continue;
+                }
+                inferPredicates.addAll(getEqualSetAndDoReplace(expr, hasCastEqualSet.calEqualSet(expr),
+                        exprPredicates));
+            }
+        }
+        return inferPredicates;
+    }
+
+    private static class Replacer extends DefaultExpressionRewriter<Map<Expression, Expression>> {
+        public static Replacer INSTANCE = new Replacer();
+
+        public Expression replace(Expression e, Map<Expression, Expression> replaceMap) {
+            return e.accept(this, replaceMap);
+        }
+
+        @Override
+        public Expression visit(Expression expr, Map<Expression, Expression> replaceMap) {
+            expr = rewriteChildren(this, expr, replaceMap);
+            if (replaceMap.containsKey(expr)) {
+                return replaceMap.get(expr);
+            }
+            return expr;
+        }
+    }
+
+    /* This function is used to input a=b b=c to derive a=c, and return a=c.*/
+    private static Set<Expression> deduceTransitiveEquality(ImmutableEqualSet<Slot> equalSet,
+            Set<Pair<Slot, Slot>> equalPairs) {
+        List<Set<Slot>> equalSetList = equalSet.calEqualSetList();
+        Set<Expression> derivedEqualities = new LinkedHashSet<>();
+        for (Set<Slot> es : equalSetList) {
+            List<Slot> el = es.stream().sorted(Comparator.comparingInt(s -> s.getExprId().asInt()))
+                    .collect(Collectors.toList());
+            for (int i = 0; i < el.size(); i++) {
+                Slot left = el.get(i);
+                for (int j = i + 1; j < el.size(); j++) {
+                    Slot right = el.get(j);
+                    if (equalPairs.contains(Pair.of(left, right))) {
+                        continue;
+                    }
+                    EqualTo newEqualTo = new EqualTo(left, right);
+                    if (isSingleTableExpression(newEqualTo)) {
+                        continue;
+                    }
+                    derivedEqualities.add(TypeCoercionUtils.processComparisonPredicate(newEqualTo)
+                            .withInferred(true));
+                }
+            }
+        }
+        return derivedEqualities;
+    }
+
+    private static boolean isSingleTableExpression(Expression expr) {
```

Comment:
maybe singleRelationExpression is a better name. add a `getQualifier` function for slot?